controllers/apps/operations/horizontal_scaling.go

```diff
@@ -447,3 +460,152 @@ func (hs horizontalScalingOpsHandler) getCompExpectedOfflineInstances(
+		}
+	}
+	if !reflect.DeepEqual(opsDeepCopy.Status, ops.Status) {
+		if err = cli.Status().Patch(reqCtx.Ctx, ops, client.MergeFrom(opsDeepCopy)); err != nil {
+			return err
+		}
+	}
+	if ops.Annotations[constant.RetryAbortedHScaleAnnoKey] == "true" {
+		ops.Annotations[constant.RetryAbortedHScaleAnnoKey] = "done"
+	}
+	if !reflect.DeepEqual(opsDeepCopy.Annotations, ops.Annotations) {
+		return cli.Patch(reqCtx.Ctx, ops, client.MergeFrom(opsDeepCopy))
+	}
+	return nil
+}
+
+func handleClusterCompSpecForAbortedHScale(opsRes *OpsResource, expectProgressStatus, toProgressStatus appsv1alpha1.ProgressStatus, reverse bool) (bool, error) {
+	needUpdateCluster := false
+	// update component and instance template replicas
+	updateCompReplicas := func(compSpec *appsv1alpha1.ClusterComponentSpec, tplName string, operation func(replicas int32) int32) {
+		compSpec.Replicas = operation(compSpec.Replicas)
+		if tplName != "" {
+			for k := range compSpec.Instances {
+				instanceTPL := &compSpec.Instances[k]
+				if instanceTPL.Name == tplName {
+					instanceTPL.Replicas = pointer.Int32(operation(instanceTPL.GetReplicas()))
+				}
+			}
+		}
+	}
+	// offline the required instances
+	offlineInstance := func(compSpec *appsv1alpha1.ClusterComponentSpec, tplName, podName string) {
+		compSpec.OfflineInstances = append(compSpec.OfflineInstances, podName)
+		updateCompReplicas(compSpec, tplName, func(replicas int32) int32 {
+			return replicas - int32(1)
+		})
+	}
+	isCreateOperation := func(progressGroup, compName string) bool {
+		return (progressGroup == fmt.Sprintf("%s/%s", compName, createPodOperation) && !reverse) ||
+			(progressGroup == fmt.Sprintf("%s/%s", compName, deletePodOperation) && reverse)
+	}
+	needReCreatedPodMap := map[string]*appsv1alpha1.ProgressStatusDetail{}
+	clusterCompMap := map[string]*appsv1alpha1.ClusterComponentSpec{}
+	for i, comp := range opsRes.Cluster.Spec.ComponentSpecs {
+		clusterCompMap[comp.Name] = &opsRes.Cluster.Spec.ComponentSpecs[i]
+	}
+	for compName := range opsRes.OpsRequest.Status.Components {
+		compSpec, ok := clusterCompMap[compName]
+		if !ok {
+			continue
+		}
+		compStatus := opsRes.OpsRequest.Status.Components[compName]
+		for j := range compStatus.ProgressDetails {
+			pd := &compStatus.ProgressDetails[j]
+			if pd.Status != expectProgressStatus {
+				continue
+			}
+			currExpectPodSet, err := intctrlcomp.GenerateAllPodNamesToSet(compSpec.Replicas, compSpec.Instances,
+				compSpec.OfflineInstances, opsRes.Cluster.Name, compName)
+			if err != nil {
+				return false, err
+			}
+			podName := strings.Replace(pd.ObjectKey, constant.PodKind+"/", "", 1)
+			tplName := appsv1alpha1.GetInstanceTemplateName(opsRes.Cluster.Name, compSpec.Name, podName)
+			if isCreateOperation(pd.Group, compName) {
+				if _, ok = currExpectPodSet[podName]; ok {
+					// offline the pods that have not been created yet during scaling out.
+					offlineInstance(compSpec, tplName, podName)
+					needUpdateCluster = true
```

Comment:
If ops.Annotations[constant.RetryAbortedHScaleAnnoKey] is true, "reverse" is true, deleted pods (aborted) will be put into offlineInstances.

If ops.Annotations[constant.RetryAbortedHScaleAnnoKey] is false, "reverse" if false, created pods (aborted) will be put into offlineInstances.

right?